aspnetcore/security/key-vault-configuration.md

```diff
@@ -0,0 +1,144 @@
+
+Assume we have several `ConnectionString` key vault secrets with the application name prefixed. For the sample application, we would create a secret in key vault for `KeyVaultConfigProviderSample-ConnectionString` and its value. For a second application, we would create a secret for `SomeOtherApplicationName-ConnectionString` and its value. We would like each app to load its own `ConnectionString` secret into its configuration as `ConnectionString`. An example of this implementation is shown below.
+
+```csharp
+public class EnvironmentSecretManager : IKeyVaultSecretManager
+{
+    private readonly string _appNamePrefix;
+
+    public EnvironmentSecretManager(string appName)
+    {
+        _appNamePrefix = appName + "-";
+    }
+
+    public bool Load(SecretItem secret)
+    {
+        return secret.Identifier.Name.StartsWith(_appNamePrefix);
+    }
+
+    public string GetKey(SecretBundle secret)
+    {
+        return secret.SecretIdentifier.Name.Substring(_appNamePrefix.Length);
+    }
+}
+```
+
+```csharp
+builder.AddAzureKeyVault(
+    $"https://{config["Vault"]}.vault.azure.net/",
+    config["ClientId"],
+    config["ClientSecret"],
+    new EnvironmentSecretManager(env.ApplicationName));
+    
+Configuration = builder.Build();
+```
+
+The `Load()` method is called by an algorithm in the provider that iterates through the secrets to find the one that matches the application name as a prefix to the secret's name. When a match is found with `Load()`, the algorithm uses the `GetKey()` method to return the configuration name of the secret name. It simply strips off the application name prefix from the secret's name and returns the name for loading into the app's configuration name-value pairs.
+
+If you implemented this approach with the sample application, the key vault secrets would be loaded, the string secret for `KeyVaultConfigProviderSample-ConnectionString` would be matched, and the application name `KeyVaultConfigProviderSample` (with the dash) would be stripped off to load `ConnectionString` with its value into the app's configuration.
+
+You can also provide your own `KeyVaultClient` implementation to `AddAzureKeyVault()`. Supplying a custom client allows you to share a single instance of the client between the configuration provider and other parts of your application.
+
+## Controlling access to the ClientSecret
+To maintain the `ClientSecret` outside of your project tree, you can use the [Secret Manager tool](xref:security/app-secrets), whereby you can associate app secrets with a specific project and share them across multiple projects. However, the Secret Manager tool does not encrypt the stored secrets and should not be treated as a trusted store.
```

Comment:
~~ However,~~ the Secret Manager tool does not encrypt the stored secrets and should not be treated as a trusted store.